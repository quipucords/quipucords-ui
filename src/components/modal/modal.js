import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { useMount, useUnmount } from 'react-use';
import { Modal as PfModal, ModalProps, ModalVariant } from '@patternfly/react-core';
import classNames from 'classnames';
import { translate } from '../i18n/i18n';

/**
 * Wrapper for adjusting PF Modal styling.
 *
 * @param {object} props
 * @param {string} props.'aria-label'
 * @param {boolean} props.backdrop
 * @param {React.ReactNode} props.children
 * @param {string} props.className
 * @param {React.ReactNode|Function} props.header
 * @param {React.ReactNode|Function} props.footer
 * @param {boolean} props.isContentOnly
 * @param {string} props.position
 * @param {string} props.positionOffset
 * @param {boolean} props.showClose
 * @param {Function} props.t
 * @param {string} props.variant
 * @param {ModalProps} props.props
 * @returns {React.ReactNode}
 */
const Modal = ({
  'aria-label': ariaLabel,
  backdrop,
  children,
  className,
  header,
  footer,
  isContentOnly,
  position,
  positionOffset,
  showClose,
  t,
  variant,
  ...props
}) => {
  const [element, setElement] = useState();
  const updatedProps = { ...props };
  const cssClassName = classNames(
    `quipucords-modal`,
    { 'quipucords-modal__hide-backdrop': backdrop === false },
    { 'quipucords-modal__rcue-width': !variant },
    className
  );

  useMount(() => {
    const domElement = document.createElement('div');
    document.body.appendChild(domElement);
    setElement(domElement);
  });

  useUnmount(() => {
    element?.remove();
  });

  if (!element) {
    return null;
  }

  element.className = cssClassName;

  if (header) {
    updatedProps.header = (typeof header === 'function' && header()) || header;
  }

  if (footer) {
    updatedProps.footer = (typeof footer === 'function' && footer()) || footer;
  }

  return (
    <PfModal
      appendTo={element}
      aria-label={ariaLabel || t('modal.aria-label-default')}
      hasNoBodyWrapper={isContentOnly}
      position={position}
      positionOffset={positionOffset}
      showClose={showClose}
      variant={variant}
      {...updatedProps}
    >
      {(React.isValidElement(children) && children) || <div>{children || ''}</div>}
    </PfModal>
  );
};

/**
 * Prop types
 *
 * @type {{backdrop: boolean, showClose: boolean, t: Function, children: React.ReactNode,
 *     footer: React.ReactNode|Function, variant: string, header: React.ReactNode|Function,
 *     className: string|object, isContentOnly: boolean, position: string, positionOffset: string,
 *     'aria-label': string}}
 */
Modal.propTypes = {
  'aria-label': PropTypes.string,
  backdrop: PropTypes.bool,
  children: PropTypes.node.isRequired,
  className: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  footer: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  header: PropTypes.oneOfType([PropTypes.func, PropTypes.node]),
  isContentOnly: PropTypes.bool,
  position: PropTypes.oneOf(['top', null]),
  positionOffset: PropTypes.string,
  showClose: PropTypes.bool,
  t: PropTypes.func,
  variant: PropTypes.oneOf([...Object.values(ModalVariant)])
};

/**
 * Default props
 *
 * @type {{backdrop: boolean, showClose: boolean, t: translate, footer: null, variant: null, header: null,
 *     className: null, isContentOnly: boolean, position: string, positionOffset: string, 'aria-label': null}}
 */
Modal.defaultProps = {
  'aria-label': null,
  backdrop: true,
  className: null,
  footer: null,
  header: null,
  isContentOnly: false,
  position: 'top',
  positionOffset: '5%',
  showClose: false,
  t: translate,
  variant: null
};

export { Modal as default, Modal, ModalVariant };
