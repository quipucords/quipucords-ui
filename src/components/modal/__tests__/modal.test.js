import React from 'react';
import { Modal as PfModal, ModalContent } from '@patternfly/react-core';
import { Modal } from '../modal';

describe('Modal Component', () => {
  it('should render a basic component', async () => {
    const props = {};

    const component = await mountHookComponent(<Modal {...props}>lorem ipsum</Modal>);
    expect(component).toMatchSnapshot('basic');
  });

  it('should allow modifying specific and custom props', async () => {
    const props = {
      backdrop: false
    };

    const backdropComponent = await mountHookComponent(<Modal {...props}>lorem ipsum</Modal>);
    expect(backdropComponent.find(PfModal)).toMatchSnapshot('backdrop');

    props.backdrop = true;
    props['aria-label'] = 'dolor sit';

    const ariaLabelComponent = await mountHookComponent(<Modal {...props}>lorem ipsum</Modal>);
    expect(ariaLabelComponent.find(PfModal)).toMatchSnapshot('aria-label');
  });

  it('should allow custom headers and footers', async () => {
    // disableFocusTrap for testing only
    const props = {
      isOpen: true,
      disableFocusTrap: true
    };

    const component = await mountHookComponent(<Modal {...props}>hello world</Modal>);

    component.setProps({
      header: undefined,
      footer: undefined
    });

    expect(component.find(ModalContent).render()).toMatchSnapshot('undefined');

    component.setProps({
      header: 'lorem ipsum',
      footer: 'dolor sit'
    });

    expect(component.find(ModalContent).render()).toMatchSnapshot('string');

    component.setProps({
      header: () => 'lorem ipsum',
      footer: () => 'dolor sit'
    });

    expect(component.find(ModalContent).render()).toMatchSnapshot('function');

    component.setProps({
      header: [<React.Fragment key="lorem">lorem ipsum</React.Fragment>],
      footer: [<React.Fragment key="dolor">dolor sit</React.Fragment>]
    });

    expect(component.find(ModalContent).render()).toMatchSnapshot('list');

    component.setProps({
      header: <React.Fragment>lorem ipsum</React.Fragment>,
      footer: <React.Fragment key="dolor">dolor sit</React.Fragment>
    });

    expect(component.find(ModalContent).render()).toMatchSnapshot('element');
  });
});
