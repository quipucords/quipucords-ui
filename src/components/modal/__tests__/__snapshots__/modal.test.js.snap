// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Component should allow custom headers and footers: element 1`] = `
<div
  class="pf-c-backdrop"
>
  <div
    class="pf-l-bullseye"
  >
    <div
      aria-describedby="pf-modal-part-6"
      aria-label="t(modal.aria-label-default)"
      aria-labelledby="pf-modal-part-4"
      aria-modal="true"
      class="pf-c-modal-box pf-m-align-top"
      data-ouia-component-id="OUIA-Generated-Modal-5"
      data-ouia-component-type="PF4/ModalContent"
      data-ouia-safe="true"
      id="pf-modal-part-4"
      role="dialog"
      style="--pf-c-modal-box--m-align-top--spacer: 5%;"
    >
      <header
        class="pf-c-modal-box__header"
      >
        lorem ipsum
      </header>
      <div
        class="pf-c-modal-box__body"
        id="pf-modal-part-6"
      >
        <div>
          hello world
        </div>
      </div>
      <footer
        class="pf-c-modal-box__footer"
      >
        dolor sit
      </footer>
    </div>
  </div>
</div>
`;

exports[`Modal Component should allow custom headers and footers: function 1`] = `
<div
  class="pf-c-backdrop"
>
  <div
    class="pf-l-bullseye"
  >
    <div
      aria-describedby="pf-modal-part-6"
      aria-label="t(modal.aria-label-default)"
      aria-labelledby="pf-modal-part-4"
      aria-modal="true"
      class="pf-c-modal-box pf-m-align-top"
      data-ouia-component-id="OUIA-Generated-Modal-5"
      data-ouia-component-type="PF4/ModalContent"
      data-ouia-safe="true"
      id="pf-modal-part-4"
      role="dialog"
      style="--pf-c-modal-box--m-align-top--spacer: 5%;"
    >
      <header
        class="pf-c-modal-box__header"
      >
        lorem ipsum
      </header>
      <div
        class="pf-c-modal-box__body"
        id="pf-modal-part-6"
      >
        <div>
          hello world
        </div>
      </div>
      <footer
        class="pf-c-modal-box__footer"
      >
        dolor sit
      </footer>
    </div>
  </div>
</div>
`;

exports[`Modal Component should allow custom headers and footers: list 1`] = `
<div
  class="pf-c-backdrop"
>
  <div
    class="pf-l-bullseye"
  >
    <div
      aria-describedby="pf-modal-part-6"
      aria-label="t(modal.aria-label-default)"
      aria-labelledby="pf-modal-part-4"
      aria-modal="true"
      class="pf-c-modal-box pf-m-align-top"
      data-ouia-component-id="OUIA-Generated-Modal-5"
      data-ouia-component-type="PF4/ModalContent"
      data-ouia-safe="true"
      id="pf-modal-part-4"
      role="dialog"
      style="--pf-c-modal-box--m-align-top--spacer: 5%;"
    >
      <header
        class="pf-c-modal-box__header"
      >
        lorem ipsum
      </header>
      <div
        class="pf-c-modal-box__body"
        id="pf-modal-part-6"
      >
        <div>
          hello world
        </div>
      </div>
      <footer
        class="pf-c-modal-box__footer"
      >
        dolor sit
      </footer>
    </div>
  </div>
</div>
`;

exports[`Modal Component should allow custom headers and footers: string 1`] = `
<div
  class="pf-c-backdrop"
>
  <div
    class="pf-l-bullseye"
  >
    <div
      aria-describedby="pf-modal-part-6"
      aria-label="t(modal.aria-label-default)"
      aria-labelledby="pf-modal-part-4"
      aria-modal="true"
      class="pf-c-modal-box pf-m-align-top"
      data-ouia-component-id="OUIA-Generated-Modal-5"
      data-ouia-component-type="PF4/ModalContent"
      data-ouia-safe="true"
      id="pf-modal-part-4"
      role="dialog"
      style="--pf-c-modal-box--m-align-top--spacer: 5%;"
    >
      <header
        class="pf-c-modal-box__header"
      >
        lorem ipsum
      </header>
      <div
        class="pf-c-modal-box__body"
        id="pf-modal-part-6"
      >
        <div>
          hello world
        </div>
      </div>
      <footer
        class="pf-c-modal-box__footer"
      >
        dolor sit
      </footer>
    </div>
  </div>
</div>
`;

exports[`Modal Component should allow custom headers and footers: undefined 1`] = `
<div
  class="pf-c-backdrop"
>
  <div
    class="pf-l-bullseye"
  >
    <div
      aria-describedby="pf-modal-part-6"
      aria-label="t(modal.aria-label-default)"
      aria-labelledby="pf-modal-part-4"
      aria-modal="true"
      class="pf-c-modal-box pf-m-align-top"
      data-ouia-component-id="OUIA-Generated-Modal-5"
      data-ouia-component-type="PF4/ModalContent"
      data-ouia-safe="true"
      id="pf-modal-part-4"
      role="dialog"
      style="--pf-c-modal-box--m-align-top--spacer: 5%;"
    >
      <div
        class="pf-c-modal-box__body"
        id="pf-modal-part-6"
      >
        <div>
          hello world
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal Component should allow modifying specific and custom props: aria-label 1`] = `
<Modal
  actions={Array []}
  appendTo={
    <div
      class="quipucords-modal quipucords-modal__rcue-width"
    >
      <div />
    </div>
  }
  aria-describedby=""
  aria-label="dolor sit"
  aria-labelledby=""
  className=""
  hasNoBodyWrapper={false}
  isOpen={false}
  onClose={[Function]}
  ouiaSafe={true}
  position="top"
  positionOffset="5%"
  showClose={false}
  title=""
  titleIconVariant={null}
  titleLabel=""
  variant={null}
>
  <Portal
    containerInfo={<div />}
  >
    <ModalContent
      actions={Array []}
      aria-describedby=""
      aria-label="dolor sit"
      aria-labelledby=""
      boxId="pf-modal-part-2"
      className=""
      descriptorId="pf-modal-part-4"
      hasNoBodyWrapper={false}
      isOpen={false}
      labelId="pf-modal-part-3"
      onClose={[Function]}
      ouiaId="OUIA-Generated-Modal-3"
      ouiaSafe={true}
      position="top"
      positionOffset="5%"
      showClose={false}
      title=""
      titleIconVariant={null}
      titleLabel=""
      variant={null}
    />
  </Portal>
</Modal>
`;

exports[`Modal Component should allow modifying specific and custom props: backdrop 1`] = `
<Modal
  actions={Array []}
  appendTo={
    <div
      class="quipucords-modal quipucords-modal__hide-backdrop quipucords-modal__rcue-width"
    >
      <div />
    </div>
  }
  aria-describedby=""
  aria-label="t(modal.aria-label-default)"
  aria-labelledby=""
  className=""
  hasNoBodyWrapper={false}
  isOpen={false}
  onClose={[Function]}
  ouiaSafe={true}
  position="top"
  positionOffset="5%"
  showClose={false}
  title=""
  titleIconVariant={null}
  titleLabel=""
  variant={null}
>
  <Portal
    containerInfo={<div />}
  >
    <ModalContent
      actions={Array []}
      aria-describedby=""
      aria-label="t(modal.aria-label-default)"
      aria-labelledby=""
      boxId="pf-modal-part-1"
      className=""
      descriptorId="pf-modal-part-3"
      hasNoBodyWrapper={false}
      isOpen={false}
      labelId="pf-modal-part-2"
      onClose={[Function]}
      ouiaId="OUIA-Generated-Modal-2"
      ouiaSafe={true}
      position="top"
      positionOffset="5%"
      showClose={false}
      title=""
      titleIconVariant={null}
      titleLabel=""
      variant={null}
    />
  </Portal>
</Modal>
`;

exports[`Modal Component should allow modifying specific and custom props: isContentOnly 1`] = `
<Modal
  actions={Array []}
  appendTo={
    <div
      class="quipucords-modal quipucords-modal__hide-backdrop quipucords-modal__rcue-width"
    >
      <div />
    </div>
  }
  aria-describedby=""
  aria-label="t(modal.aria-label-default)"
  aria-labelledby=""
  className=""
  hasNoBodyWrapper={true}
  isOpen={false}
  onClose={[Function]}
  ouiaSafe={true}
  position="top"
  positionOffset="5%"
  showClose={false}
  title=""
  titleIconVariant={null}
  titleLabel=""
  variant={null}
>
  <Portal
    containerInfo={<div />}
  >
    <ModalContent
      actions={Array []}
      aria-describedby=""
      aria-label="t(modal.aria-label-default)"
      aria-labelledby=""
      boxId="pf-modal-part-3"
      className=""
      descriptorId="pf-modal-part-5"
      hasNoBodyWrapper={true}
      isOpen={false}
      labelId="pf-modal-part-4"
      onClose={[Function]}
      ouiaId="OUIA-Generated-Modal-4"
      ouiaSafe={true}
      position="top"
      positionOffset="5%"
      showClose={false}
      title=""
      titleIconVariant={null}
      titleLabel=""
      variant={null}
    />
  </Portal>
</Modal>
`;

exports[`Modal Component should render a basic component: basic 1`] = `
<Modal
  aria-label={null}
  backdrop={true}
  className={null}
  footer={null}
  header={null}
  isContentOnly={false}
  position="top"
  positionOffset="5%"
  showClose={false}
  t={[Function]}
  variant={null}
>
  <Modal
    actions={Array []}
    appendTo={
      <div
        class="quipucords-modal quipucords-modal__rcue-width"
      >
        <div />
      </div>
    }
    aria-describedby=""
    aria-label="t(modal.aria-label-default)"
    aria-labelledby=""
    className=""
    hasNoBodyWrapper={false}
    isOpen={false}
    onClose={[Function]}
    ouiaSafe={true}
    position="top"
    positionOffset="5%"
    showClose={false}
    title=""
    titleIconVariant={null}
    titleLabel=""
    variant={null}
  >
    <Portal
      containerInfo={<div />}
    >
      <ModalContent
        actions={Array []}
        aria-describedby=""
        aria-label="t(modal.aria-label-default)"
        aria-labelledby=""
        boxId="pf-modal-part-0"
        className=""
        descriptorId="pf-modal-part-2"
        hasNoBodyWrapper={false}
        isOpen={false}
        labelId="pf-modal-part-1"
        onClose={[Function]}
        ouiaId="OUIA-Generated-Modal-1"
        ouiaSafe={true}
        position="top"
        positionOffset="5%"
        showClose={false}
        title=""
        titleIconVariant={null}
        titleLabel=""
        variant={null}
      />
    </Portal>
  </Modal>
</Modal>
`;
